minetest/minetest
# Show relative screenshot filename on chat when applicable

This PR allows printing the relative filename of screenshot to client instead of absolute when screenshot_path is relative. When screenshot path is absolute it prints absolute path like before.

- Fix the issue where absolute filename with dots (e.g. `/path/to/minetest/bin/../screenshots/filename.ext`) is displayed in chat when screenshot path is relative (e.g. `screenshots`)
- The files listed below have all the necessary changes and modifications implementing such functions are listed near each file:
  - `src/client/client.cpp`, `Client::makeScreenshot()`:
    - Upon saving screenshot relative path is generated using following method
      - Initially, output text is set to filename, with relative components removed by using function `fs::AbsolutePath`
      - Screenshot dir then is checked if it is root dir, value is compared to output of `fs::GetRootDir`, If it is not the same, the operations listed below are performed to get the relative file name
        - Common branching path of minetest share path and absolute screenshot path is generated using new function `fs::GetCommonPath`, dfference of levels is stored in the variable pointing to third argument of that function
        - If screenshot path is not absolute and level difference is lover than set threshold, the relative path generated using a new function `fs::BuildRelativePath` is used inplace, otherwise no changes are done.
      - If screenshot dir is root, no changes are done to output filename
  - `src/filesys.h` & `src/filesys.cpp`
    - `RemoveRelativePathComponents`
      - Upon return, the resulting path checked if it is not same as root dir, trailing character is not removed if path is the root dir.
    - `AbsolutePath`
      - unix logic is changed:
        - If path is absolute, `RemoveRelativePathComponents` is used to return clean path
        - Otherwise current working dir is get using `getCurrentWorkingDir`, if it fails the function returns same value as above, otherwise it returns working dir + path
      - Suggestions welcome, this function might not be currently finalized atm.
    - New function `SplitComponentFromPath`, splits last component from path into variable supplied to second argument, also updating variable supplied to first argument.
      - Returns true if path contains delimiter, component is top file/directory in the path.
      - Otherwise returns false, component is same as the path.
    - New function `GetRootDir`, returns root directory, on windows returning notation like `C:\`, etc.
    - New function `GetCommonPath`, compares two paths, returns a path to the branching point of first and second argument, third argument contains returned level difference between them. 
  - `src/porting.h`
    - added missing definition of `porting::getCurrentWorkingDir`, used by some functions in `filesys.cpp`
- Those changes make sure path is shown w/o upper level markers and single dots anywhere in the middle of the path except for upper level markers on beginning of the relative path where needed. Also it makes sure no extra slash is printed when path is empty too or is show as root (`/`)
- Any suggestions or bug reports welcome, most features work good so far.

## To do

This PR is a Work in Progress ~Ready for Review~.

  - [x] Make sure absolute screenshot dir is derived properly
  - [x] If absolute value of screenshot dir is on drive or system root, displayed (relative) path should change to absolute
  - [x] When screenshot dir is not on system root
    - [x] Absolute path for minetest directory should be provided by `fs::AbsolutePath`
    - [x] Common path & level count from minetest share dir to it should be provided by new function `fs::GetCommonPath`, level difference should be stored in the variable supplied to the third argument of this function
   - [x] If screenshot path config variable were absolute or level difference were too high screenshot dir should be set to be absolute 
      - [x] Otherwise relative path should be built using function `fs::BuildRelativePath`
  - [x] Basename should have no information prepended
  - [x] Relative and absolute filenames should consist of basename and path to screenshots, joined using new function `fs::JoinPaths`
  - [x] Absolute path should be used to create screenshot dir and for screenshot file saving
  - [ ] Make sure the new functions are efficient, no bugs, etc.
  - [ ] Make sure everything else works, might need some testing (messages in `infostream` would show if anything would be broken)
  - [ ] Remove debug messages when done

## How to test

- Build minetest from master, take screenshot, notice both relative and absolute screenshot paths print as absolute, parent directory markers (`../`) and dots (`./`) not removed or substituted with alternative path
- Build minetest from PR, do tests by setting screenshot path as shown below, files should save in paths relative to share or minetest dir, depending on build
  - default screenshot path - displayed file name should have a default sub-directory preceding it, no markers shown before it
  - path with multiple ascending and descending levels should resolve to shorter and direct path, e.g `../something/../minetest`  should resolve to `minetest` and `../something/..` should resolve to `..`, w/o delimiter after it
  - empty screenshot path - only file name should be displayed, no markers before it
  - path pointing to parent directory - displayed file name should include correct count of parent level markers depending on parent level depth
  - path descending with more parent levels then set limit (currently 3) - displayed file name should always be absolute
  - any absolute path - displayed file name should be absolute, any kind of unneeded markers and extra delimiters should be removed
